File: master_plan_it/master_plan_it/report/mpit_economic_detail/mpit_economic_detail.py
from __future__ import annotations

import frappe
from frappe import _

from master_plan_it.master_plan_it.financial_engine import get_economic_detail_dataset


def execute(filters=None):
    filters = frappe._dict(filters or {})
    dataset = get_economic_detail_dataset(filters)
    return _get_columns(filters), dataset.get("rows", []), None, dataset.get("chart"), _get_report_summary(dataset.get("summary", {}))


def _get_columns(filters) -> list[dict]:
    return [
        {"label": _("Date"), "fieldname": "effective_date", "fieldtype": "Date", "width": 100},
        {"label": _("Cost Center"), "fieldname": "cost_center", "fieldtype": "Link", "options": "MPIT Cost Center", "width": 140},
        {"label": _("Type"), "fieldname": "source_type", "fieldtype": "Data", "width": 90},
        {"label": _("Document"), "fieldname": "source_document", "fieldtype": "Data", "width": 170},
        {"label": _("Row"), "fieldname": "source_row", "fieldtype": "Data", "width": 160},
        {"label": _("Description"), "fieldname": "description", "fieldtype": "Data", "width": 220},
        {"label": _("Vendor"), "fieldname": "vendor", "fieldtype": "Link", "options": "MPIT Vendor", "width": 140},
        {"label": _("Project"), "fieldname": "project", "fieldtype": "Link", "options": "MPIT Project", "width": 140},
        {"label": _("Project Stage"), "fieldname": "project_stage", "fieldtype": "Data", "width": 120},
        {"label": _("Contract"), "fieldname": "contract", "fieldtype": "Link", "options": "MPIT Contract", "width": 140},
        {"label": _("Phase"), "fieldname": "row_phase", "fieldtype": "Data", "width": 90},
        {"label": _("Funding"), "fieldname": "funding", "fieldtype": "Data", "width": 100},
        {"label": _("Net"), "fieldname": "amount_net", "fieldtype": "Currency", "width": 120},
        {"label": _("VAT"), "fieldname": "amount_vat", "fieldtype": "Currency", "width": 120},
        {"label": _("Gross"), "fieldname": "amount_gross", "fieldtype": "Currency", "width": 120},
        {"label": _("Row State"), "fieldname": "row_state", "fieldtype": "Data", "width": 100},
    ]


def _get_report_summary(summary: dict) -> list[dict]:
    return [
        {"label": _("Rows"), "value": summary.get("total_rows", 0), "datatype": "Int", "indicator": "Grey"},
        {"label": _("Net"), "value": summary.get("amount_net", 0), "datatype": "Currency", "indicator": "Blue"},
        {"label": _("VAT"), "value": summary.get("amount_vat", 0), "datatype": "Currency", "indicator": "Orange"},
        {"label": _("Gross"), "value": summary.get("amount_gross", 0), "datatype": "Currency", "indicator": "Green"},
    ]

File: master_plan_it/master_plan_it/report/mpit_economic_detail/mpit_economic_detail.js
frappe.query_reports["MPIT Economic Detail"] = {
    filters: [
        { fieldname: "year", label: __("Year"), fieldtype: "Link", options: "MPIT Year", default: String(new Date().getFullYear()), reqd: 1 },
        { fieldname: "cost_center", label: __("Cost Center"), fieldtype: "Link", options: "MPIT Cost Center" },
        { fieldname: "include_children", label: __("Include Children"), fieldtype: "Check", default: 1, depends_on: "eval:doc.cost_center" },
        { fieldname: "project", label: __("Project"), fieldtype: "Link", options: "MPIT Project" },
        { fieldname: "contract", label: __("Contract"), fieldtype: "Link", options: "MPIT Contract" },
        { fieldname: "vendor", label: __("Vendor"), fieldtype: "Link", options: "MPIT Vendor" },
        { fieldname: "expense_kind", label: __("Expense Kind"), fieldtype: "Select", options: "All\nOrdinary\nPlafond", default: "All" },
        { fieldname: "row_phase", label: __("Row Phase"), fieldtype: "Select", options: "All\nEstimate\nQuote\nActual", default: "All" },
        { fieldname: "funding", label: __("Funding"), fieldtype: "Select", options: "All\nStandard\nPlafond\nExtra", default: "All" },
        { fieldname: "project_stage", label: __("Project Stage"), fieldtype: "Select", options: "All\nIdea\nProposed\nApproved\nDeferred\nRejected\nNo Project", default: "All" },
        { fieldname: "date_basis", label: __("Date Basis"), fieldtype: "Select", options: "Spend Date\nStart Date\nEnd Date\nEffective Period", default: "Spend Date" },
        { fieldname: "from_date", label: __("From Date"), fieldtype: "Date" },
        { fieldname: "to_date", label: __("To Date"), fieldtype: "Date" },
        { fieldname: "group_by", label: __("Group By"), fieldtype: "Select", options: "Cost Center", default: "Cost Center" },
        { fieldname: "show_replaced_cancelled", label: __("Show Replaced/Cancelled"), fieldtype: "Check", default: 0 },
        { fieldname: "print_profile", label: __("Print Profile"), fieldtype: "Select", options: "Executive\nStandard\nDetailed\nAudit", default: "Standard" },
        { fieldname: "print_orientation", label: __("Print Orientation"), fieldtype: "Select", options: "Auto\nPortrait\nLandscape", default: "Auto" },
        { fieldname: "print_density", label: __("Print Density"), fieldtype: "Select", options: "Normal\nCompact", default: "Normal" },
    ],

    formatter(value, row, column, data, default_formatter) {
        const formatted = default_formatter(value, row, column, data);
        if (column.fieldname === "status" && value === "Critical") {
            return `<span class="text-danger fw-semibold">${formatted}</span>`;
        }
        if (column.fieldname === "status" && value === "Warning") {
            return `<span class="text-warning fw-semibold">${formatted}</span>`;
        }
        if (column.fieldname === "recommended_action" && ["Expired", "Renegotiate"].includes(value)) {
            return `<span class="text-danger fw-semibold">${formatted}</span>`;
        }
        return formatted;
    },
};
